openshift-kube-apiserver/admission/autoscaling/managednode/admission.go

```diff
@@ -0,0 +1,135 @@
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/kubernetes/pkg/kubelet/managed"
+
+	corev1 "k8s.io/api/core/v1"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	PluginName = "autoscaling.openshift.io/ManagedNode"
+	// infraClusterName contains the name of the cluster infrastructure resource
+	infraClusterName = "cluster"
+)
+
+var _ = initializer.WantsExternalKubeClientSet(&managedNodeValidate{})
+var _ = admission.ValidationInterface(&managedNodeValidate{})
+var _ = WantsInfraInformer(&managedNodeValidate{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &managedNodeValidate{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+			}, nil
+		})
+}
+
+type managedNodeValidate struct {
+	*admission.Handler
+	client                kubernetes.Interface
+	infraConfigLister     configv1listers.InfrastructureLister
+	infraConfigListSynced func() bool
+}
+
+// SetExternalKubeClientSet implements the WantsExternalKubeClientSet interface.
+func (a *managedNodeValidate) SetExternalKubeClientSet(client kubernetes.Interface) {
+	a.client = client
+}
+
+func (a *managedNodeValidate) SetInfraInformer(informer configv1informer.InfrastructureInformer) {
+	a.infraConfigLister = informer.Lister()
+	a.infraConfigListSynced = informer.Informer().HasSynced
+}
+
+func (a *managedNodeValidate) ValidateInitialization() error {
+	if a.client == nil {
+		return fmt.Errorf("%s plugin needs a kubernetes client", PluginName)
+	}
+	if a.infraConfigLister == nil {
+		return fmt.Errorf("%s did not get a config infrastructure lister", PluginName)
+	}
+	if a.infraConfigListSynced == nil {
+		return fmt.Errorf("%s plugin needs a config infrastructure lister synced", PluginName)
+	}
+	return nil
+}
+
+func (a *managedNodeValidate) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if attr.GetResource().GroupResource() != corev1.Resource("nodes") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	node, ok := attr.GetObject().(*coreapi.Node)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetResource()))
+	}
+
+	// infraConfigListSynced is expected to be thread-safe since the underlying call is to the standard
+	// informer HasSynced() function which is thread-safe.
+	if !a.infraConfigListSynced() {
+		return admission.NewForbidden(attr, fmt.Errorf("%s infra config cache not synchronized", PluginName))
+	}
+
+	clusterInfra, err := a.infraConfigLister.Get(infraClusterName)
+	if err != nil {
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	// Check if we are in CPU Partitioning mode for AllNodes
+	allErrs := validateClusterCPUPartitioning(clusterInfra.Status, node)
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return admission.NewForbidden(attr, allErrs.ToAggregate())
```

Comment:
Nit, I think `admission.NewInvalid` is more appropriate, since the node is missing resource information for CPU partitioning, but once fixed it should work, right?

Reply:
Yup, that's correct, updated the code to use NewInvalid, thanks!